File: src/services/renderers/smoothSpectrum.ts
import { RenderContext } from './types';

export function renderSmoothSpectrum(ctx: RenderContext) {
  const { ctx: c, width, height, config, freqData } = ctx;
  const barCount = config.reactivity.barCount;
  const sensitivity = config.reactivity.sensitivity;
  const theme = config.theme;

  const availableWidth = width * 0.92;
  const startX = (width - availableWidth) / 2;
  const step = Math.max(1, Math.floor(freqData.length / barCount));
  const bottomY = height * 0.85;
  const maxH = height * 0.65;

  if (barCount < 2) return;

  const points: { x: number; y: number }[] = [];
  const xStep = availableWidth / (barCount - 1);

  for (let i = 0; i < barCount; i++) {
    let val = 0;
    for (let j = 0; j < step; j++) {
      val += freqData[i * step + j] || 0;
    }
    val = (val / step / 255) * sensitivity;
    val = Math.min(1, Math.max(0, val));
    const barH = val * maxH;
    const x = startX + i * xStep;
    points.push({ x, y: bottomY - barH });
  }

  const mirror = config.reactivity.mirrorBars;

  c.save();
  c.beginPath();
  c.moveTo(points[0].x, bottomY);
  for (let i = 0; i < points.length - 1; i++) {
    const xc = (points[i].x + points[i + 1].x) / 2;
    const yc = (points[i].y + points[i + 1].y) / 2;
    c.quadraticCurveTo(points[i].x, points[i].y, xc, yc);
  }
  const last = points[points.length - 1];
  c.lineTo(last.x, last.y);
  c.lineTo(last.x, bottomY);
  c.closePath();

  const fillGrad = c.createLinearGradient(0, bottomY - maxH, 0, bottomY);
  fillGrad.addColorStop(0, theme.primaryColor);
  fillGrad.addColorStop(0.5, theme.secondaryColor);
  fillGrad.addColorStop(1, 'transparent');
  c.fillStyle = fillGrad;
  c.shadowBlur = 20;
  c.shadowColor = theme.glowColor;
  c.fill();

  if (mirror) {
    c.globalAlpha = 0.5;
    c.beginPath();
    c.moveTo(points[0].x, bottomY);
    for (let i = 0; i < points.length - 1; i++) {
      const xc = (points[i].x + points[i + 1].x) / 2;
      const yc = bottomY + (bottomY - (points[i].y + points[i + 1].y) / 2);
      const my = bottomY + (bottomY - points[i].y);
      c.quadraticCurveTo(points[i].x, my, xc, yc);
    }
    const myLast = bottomY + (bottomY - last.y);
    c.lineTo(last.x, myLast);
    c.lineTo(last.x, bottomY);
    c.closePath();
    c.fill();
    c.globalAlpha = 1;
  }

  c.restore();

  c.save();
  c.beginPath();
  c.moveTo(points[0].x, points[0].y);
  for (let i = 0; i < points.length - 1; i++) {
    const xc = (points[i].x + points[i + 1].x) / 2;
    const yc = (points[i].y + points[i + 1].y) / 2;
    c.quadraticCurveTo(points[i].x, points[i].y, xc, yc);
  }
  c.lineTo(last.x, last.y);

  if (mirror) {
    c.moveTo(points[0].x, bottomY + (bottomY - points[0].y));
    for (let i = 0; i < points.length - 1; i++) {
      const xc = (points[i].x + points[i + 1].x) / 2;
      const yc = bottomY + (bottomY - (points[i].y + points[i + 1].y) / 2);
      const my = bottomY + (bottomY - points[i].y);
      c.quadraticCurveTo(points[i].x, my, xc, yc);
    }
    const myLast = bottomY + (bottomY - last.y);
    c.lineTo(last.x, myLast);
  }

  c.strokeStyle = theme.accentColor;
  c.lineWidth = 2;
  c.shadowBlur = 10;
  c.shadowColor = theme.glowColor;
  c.stroke();
  c.restore();
}

File: src-tauri/src/renderers/styles/smooth_spectrum.rs
//! Smooth Spectrum style renderer (`smoothSpectrum`).

use crate::gpu2d::{Color, Fill, GpuCanvas};
use crate::renderers::{
  bin_value, theme_accent, theme_glow, theme_primary, theme_secondary, RenderContext,
};

pub fn render(c: &mut GpuCanvas, ctx: &mut RenderContext) {
  let theme = &ctx.config.theme;
  let bar_count = ctx.config.reactivity.bar_count;
  let sensitivity = ctx.config.reactivity.sensitivity;
  if bar_count < 2 {
    return;
  }

  let available_width = ctx.width * 0.92;
  let start_x = (ctx.width - available_width) / 2.0;
  let step = ((ctx.freq_data.len() as f32) / bar_count as f32).floor().max(1.0) as usize;
  let bottom_y = ctx.height * 0.85;
  let max_h = ctx.height * 0.65;
  let x_step = available_width / (bar_count as f32 - 1.0);

  let mut points: Vec<(f32, f32)> = Vec::with_capacity(bar_count);
  for i in 0..bar_count {
    let val = (bin_value(ctx.freq_data, step, i) * sensitivity).clamp(0.0, 1.0);
    let bar_h = val * max_h;
    points.push((start_x + i as f32 * x_step, bottom_y - bar_h));
  }

  let mut curve: Vec<(f32, f32)> = Vec::new();
  for i in 0..points.len() - 1 {
    let (px, py) = points[i];
    let (nx, ny) = points[i + 1];
    let (cx, cy) = ((px + nx) / 2.0, (py + ny) / 2.0);
    let seg = GpuCanvas::sample_quadratic(
      if curve.is_empty() { (px, py) } else { *curve.last().unwrap() },
      (px, py),
      (cx, cy),
      6,
    );
    if curve.is_empty() {
      curve.extend(seg);
    } else {
      curve.extend(seg.into_iter().skip(1));
    }
  }
  let last = points[points.len() - 1];
  curve.push(last);

  let mut poly: Vec<(f32, f32)> = Vec::with_capacity(curve.len() + 3);
  poly.push((points[0].0, bottom_y));
  poly.extend_from_slice(&curve);
  poly.push((last.0, bottom_y));

  let fill_grad = Fill::linear_gradient(0.0, bottom_y - max_h, 0.0, bottom_y, &[
    (0.0, theme_primary(theme)),
    (0.5, theme_secondary(theme)),
    (1.0, Color::TRANSPARENT),
  ]);
  c.save();
  c.set_fill(fill_grad);
  c.set_shadow(theme_glow(theme), 20.0);
  c.fill_polygon(&poly);
  c.restore();

  c.save();
  c.set_stroke(Fill::Solid(theme_accent(theme)));
  c.set_line_width(2.0);
  c.set_shadow(theme_glow(theme), 10.0);
  c.stroke_polyline(&curve);
  c.restore();
}
